providers: fetch instance types for all providers

zonggen opened the pull request:
Fetch instance-type from cloud providers.

Closes: https://github.com/coreos/afterburn/issues/277
Signed-off-by: Zonggen (Allen) Bai <abai@redhat.com>

Commit:
aws/mod.rs: fetch instance-type for aws

Adds instances-type to the result hashmap.

Reference: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-data-categories
Signed-off-by: Zonggen Bai <abai@redhat.com>

Commit:
providers/azure: fetch the sku as instance type

Fetches sku as instance type information in Azure.

Reference: https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service
Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/gcp: fetch machine-type as instance type

Reference: https://cloud.google.com/compute/docs/storing-retrieving-metadata
Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/openstack: fetch instance-type from openstack

Reference: https://docs.openstack.org/nova/latest/user/metadata.html
Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/packet: fetch plan as intance type

Reference: https://support.packet.com/kb/articles/metadata
Signed-off-by: Allen Bai <abai@redhat.com>

lucab commented on src/providers/azure/mod.rs:
```diff
@@ -418,12 +418,28 @@ impl Azure {
         Ok(name)
     }
+
+    fn fetch_sku (&self) -> Result<Option<String>> {
```
Why is there an inner `Option<T>` here? Can the SKU be missing/empty?
I think we can either return a plain `Result<T>` (if the SKU should always be present) or an `Option<T>` (if the SKU can be missing/empty).

zonggen replied:
The logic I took is the same as `fetch_host` on line# 407.
From my understanding, the return type is coming from https://github.com/coreos/afterburn/blob/59d3c56ccefe093ec98d5aecb2579b88b7b5e9b5/src/retry/client.rs#L195 which also depends on https://github.com/coreos/afterburn/blob/59d3c56ccefe093ec98d5aecb2579b88b7b5e9b5/src/retry/client.rs#L238 the request could fail so the outer `Result<_>` and if response is 404 then return `Ok<None>` hence the inner `Option<String>`.. Am I missing something?

lucab replied:
No, you are not missing anything. But the client API you are quoting is a generic HTTP client with retries, which also supports pages with empty bodies (thus the `Result<Option<T>>` return type).
In this specific case, I think you can map the result to a simpler `Result<String>`.

zonggen replied:
I see, thanks for explaining!

lucab commented on src/providers/azure/mod.rs:
```diff
@@ -418,12 +418,28 @@ impl Azure {
     }
+
+    fn fetch_sku (&self) -> Result<Option<String>> {
+        const SKU_URL: &str = "metadata/instance/compute/sku?api-version=2017-08-01&format=text";
```
For your task, I think you want to use the `vmSize` instead of the `sku` here.

zonggen replied:
Ah, yup I think you'are right, for future reference:
 - [vmSize_example](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/sizes)
 - [sku_example](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/cli-ps-findimage#table-of-commonly-used-windows-images)

Commit:
providers/azure: Use vmSize instead of sku

Addresses improvments suggested in the review and adds
documentation in container-linux-legacy.md.

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/azure: Add mock-test for vmSize fetching

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/gcp: Add mock-test for attributes fetching

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/aws: Add mock-test for attributes fetching

Signed-off-by: Allen Bai <abai@redhat.com>

Commit:
providers/packet: Add mock-test for attributes fetching

Signed-off-by: Allen Bai <abai@redhat.com>

lucab commented on src/providers/aws/mock_tests.rs:
```diff
@@ -1,6 +1,13 @@
+use crate::providers::MetadataProvider;
+
+#[cfg(not(feature = "cl-legacy"))]
+static ENV_PREFIX: &str = "AWS";
```
I think (but I didn't try) that you can directly use those from `crate::providers::aws`. If that doesn't directly work, additionally marking them as `pub(crate)` probably would help.

zonggen replied:
It works without `pub(crate)`.
I think it works because they are from the same module and variables in `mod.rs` is visible to `mock-tests.rs`

lucab commented on src/providers/aws/mock_tests.rs:
```diff
@@ -25,3 +32,85 @@ fn test_aws_basic() {
+
+    let v = provider.attributes().unwrap();
+
+    assert_eq!(v, attributes);
```
This is a very extensive test, thanks for that!

zonggen replied:
No problem! :+1:

lucab commented on src/providers/azure/mod.rs:
```diff
@@ -17,7 +17,9 @@
-use std::net::{IpAddr, SocketAddr};
+use std::net::IpAddr;
+#[cfg(not(test))]
+use std::net::SocketAddr;
```
As it is used only once, you could also directly move this inside the relevant `get_attributes`. As you prefer.

zonggen replied:
Sure, will update to see how it looks.

lucab commented on src/providers/gcp/mock_tests.rs:
```diff
@@ -1,7 +1,13 @@
 use mockito;
 
+#[cfg(not(feature = "cl-legacy"))]
+static ENV_PREFIX: &str = "GCP";
```
Same comment as the `aws` one.

lucab commented on src/providers/packet/mock_tests.rs:
```diff
@@ -34,3 +35,41 @@ fn test_boot_checkin() {
+
+#[test]
+fn test_packet_attributes() {
+    let metadata = "{
```
You can define a raw string as `let metadata = r#"YOUR_CONTENT"#;` so that you don't have to manually escape inner quotes.

zonggen replied:
:+1: Lesson learnt..
side note: manually escaping quotes in json file is not fun :(

Commit:
providers/tests: address review comments and cleanup tests setup

Signed-off-by: Allen Bai <abai@redhat.com>

lucab commented on src/providers/packet/mock_tests.rs:
```diff
@@ -34,3 +34,42 @@ fn test_boot_checkin() {
+    let v = provider.attributes().unwrap();
+
+    assert_eq!(v, attributes);
+}
```
Let's throw a `mockito::reset()` at the bottom here, and a `try_new` + `attributes().unwrap_err()` too.

zonggen replied:
`try_new()` and `unwrap_err()` might be sufficient since in packet we are fetching the data during `try_new`, and `attributes()` would just read the struct fields.

lucab replied:
Ack (that behavior is quite terrible but not in scope for this PR).

lucab commented on src/providers/packet/mock_tests.rs:
```diff
@@ -34,3 +34,42 @@ fn test_boot_checkin() {
+    let plan = "test-plan";
+
+    let attributes = maplit::hashmap! {
+        format!("PACKET_HOSTNAME") => String::from(hostname),
```
You can use `xyz.to_string()` on both sides of the arrow (here and below).

zonggen replied:
:+1: 
Updating!

lucab commented on src/providers/packet/mock_tests.rs:
```diff
@@ -34,3 +34,42 @@ fn test_boot_checkin() {
+
+    let hostname = "test-hostname";
+    let phone_home_url = "test-url";
+    let plan = "test-plan";
```
These three are only used within the `hashmap!`, let's directly move them inline there.

zonggen replied:
Sure

lucab commented on src/providers/azure/mock_tests.rs:
```diff
@@ -113,3 +113,28 @@ fn test_hostname() {
+    m_vmsize.assert();
+    let vmsize = r.unwrap();
+    assert_eq!(vmsize, testvmsize);
+}
```
Let's throw a `mockito::reset()` here at the bottom.

Commit:
providers/mock-tests: add mockito::reset and minor changes

Signed-off-by: Allen Bai <abai@redhat.com>

lucab reviewed (approved):
LGTM!
